### PR TITLE
Add some data persistence to fall back on when getting a null response.

### PR DESCRIPTION
**Pull request by medlir**
Store the latest good data for each location that's looked up, add an age value so we know when it's from, and if we get a null response from Yahoo, fall back on that data if it's not too old.  If it is too old, fall back again on calling clear() to populate all the values with the empty defaults.  Too old value arbitrarily set to 2 hours.  Could make it a configurable setting.  Better solution would probably be to refactor the retry code in forecast itself so it could retry several times on null results response... but this seemed like an easier fix.

### CODE REVIEW

File: default.py [hunk 1]
```diff
@@ -10,6 +11,7 @@
 ADDONVERSION = ADDON.getAddonInfo('version')
 CWD          = ADDON.getAddonInfo('path').decode("utf-8")
 RESOURCE     = xbmc.translatePath( os.path.join( CWD, 'resources', 'lib' ).encode("utf-8") ).decode("utf-8")
+DATAPATH     = os.path.join(xbmc.translatePath('special://profile/').decode('utf-8'), 'addon_data', ADDONID)
```
Comment (ronie) [inline default.py]: there's no need to construct the path yourself. it is provided by xbmcaddon:
`DATAPATH = xbmc.translatePath(ADDON.getAddonInfo('profile')).decode('utf-8')`

File: default.py [hunk 2]
```diff
@@ -37,7 +37,7 @@ def refresh_locations():
         loc_name = ADDON.getSetting('Location%s' % count)
         if loc_name:
             locations += 1
-        set_property('Location%s' % count, loc_name)
+            set_property('Location%s' % count, loc_name)
```
Comment (ronie) [inline default.py]: heh, i wrote the documentation in your commit message.. and it's not 100% accurate :-)

you _do_ want to set an empty name for unconfigured locations.
although it's only useful in a few specific cases:
1) user switches from weather addon A with 4 configured locations to weather addon B with 1 configured location.
now we need to clear the properties for those locations set by the previous weather addon
2) user may have removed one or more configured locations. in this case we need to clear those properties as well.

tl;dr: no change needed here

File: default.py [hunk 3]
```diff
@@ -145,12 +147,20 @@ def clear():
...
-    wind = ''
-    atmosphere = ''
+def properties(response, loc, locid):
+    data = ''
+    
     if response and response.get('query',None) and response['query'].get('results',None) and response['query']['results'].get('channel',None):
         data = response['query']['results']['channel']
+        pickle.dump(data, open(os.path.join(DATAPATH, 'Location' + locid + '.dat'), 'wb'), protocol=0)
+    else:
+        data = pickle.load(open(os.path.join(DATAPATH, 'Location' + locid + '.dat'), 'r'))
```
Comment (ronie) [inline default.py]: i'm not familiar with cPickle, but it likely does not support vfs paths?

although they shouldn't, some users use a path-replacement on their addon_data folder,
so for reading/writing data, i'd prefer to use xbmcvfs.

**Comment by ronie** [comment]
besides my few minor comments, this looks nice in general.
will pull if you are willing to address the issues :-)

thx!

**Comment by ronie** [comment]
thanx for the fixes!

i'll most likely replace the cPickle dependency with xbmcvfs.

**Comment by medlir** [comment]
Sorry for not getting around to that part myself!  I wasn't familiar with xbmcvfs, and then before I got a chance to look into how to use it, I ended up in the hospital with appendicitis last Wednesday.  -_-

**Comment by ronie** [comment]
no worries, i'm sure being pampered by some cute nurses for a couple of days is way more fun than trying to wrap your head around some dull python methods. ;-)
